File: VECTOAux/VectoAuxiliaries/Hvac/Bus.vb
Namespace Hvac

    Public Class Bus
        Implements IBus

        'Private Property Backing
        Private _id As Integer
        Private _model As String
        Private _floorType As String
        Private _engineType As String
        Private _lengthInMetres As String
        Private _widthInMetres As String
        Private _heightInMetres As String
        Private _registeredPassengers As Integer
        Private _isDoubleDecker As Boolean

        Public ReadOnly Property Id As Integer Implements IBus.Id
            Get
                Return _id
            End Get
        End Property

        Public Property Model As String Implements IBus.Model
            Get
                Return _model
            End Get
            Set(value As String)
                If Not ModelOK(value) Then Throw New ArgumentException("Model argument is invalid")
                _model = value
            End Set
        End Property

        Public Property FloorType As String Implements IBus.FloorType
            Get
                Return _floorType
            End Get
            Set(value As String)
                If Not FloorTypeOK(value) Then Throw New ArgumentException("Model argument is invalid")
                _floorType = value
            End Set
        End Property

        Public Property EngineType As String Implements IBus.EngineType
            Get
                Return _engineType
            End Get
            Set(value As String)
                If Not EngineOK(value) Then Throw New ArgumentException("EngineType argument is invalid")
                _engineType = value
            End Set
        End Property

        Public Property LengthInMetres As Double Implements IBus.LengthInMetres
            Get
                Return _lengthInMetres
            End Get
            Set(value As Double)
                If Not DimensionOK(value) Then Throw New ArgumentException("Invalid Length")
                _lengthInMetres = value
            End Set
        End Property

        Public Property WidthInMetres As Double Implements IBus.WidthInMetres
            Get
                Return _widthInMetres
            End Get
            Set(value As Double)
                If Not DimensionOK(value) Then Throw New ArgumentException("Invalid Width")
                _widthInMetres = value
            End Set
        End Property

        Public Property HeightInMetres As Double Implements IBus.HeightInMetres
            Get
                Return _heightInMetres
            End Get
            Set(value As Double)
                If Not DimensionOK(value) Then Throw New ArgumentException("Invalid Height")
                _heightInMetres = value
            End Set
        End Property

        Public Property RegisteredPassengers As Integer Implements IBus.RegisteredPassengers
            Get
                Return _registeredPassengers
            End Get
            Set(value As Integer)
                If Not PassengersOK(value) Then Throw New ArgumentException("Invalid Number Of Passengers")
                _registeredPassengers = value
            End Set
        End Property

        Public Property IsDoubleDecker As Boolean Implements IBus.IsDoubleDecker
            Get
                Return _isDoubleDecker
            End Get
            Set(ByVal value As Boolean)
                _isDoubleDecker = value
            End Set
        End Property

        'Constructors
        Public Sub New(_id As Integer,
                        _model As String,
                        _floorType As String,
                        _engineType As String,
                        _lengthInMetres As String,
                        _widthInMetres As String,
                        _heightInMetres As String,
                        _registeredPassengers As Integer,
                        _doubleDecker As Boolean)


            'Validity checks.
            If Not ModelOK(_model) Then Throw New ArgumentException("Model argument is invalid")
            If Not FloorTypeOK(_floorType) Then Throw New ArgumentException("Model argument is invalid")
            If Not EngineOK(_engineType) Then Throw New ArgumentException("EngineType argument is invalid")
            If Not DimensionOK(_lengthInMetres) Then Throw New ArgumentException("Invalid Length")
            If Not DimensionOK(_widthInMetres) Then Throw New ArgumentException("Invalid Width")
            If Not DimensionOK(_heightInMetres) Then Throw New ArgumentException("Invalid Height")
            If Not PassengersOK(_registeredPassengers) Then Throw New ArgumentException("Invalid Number Of Passengers")


            'Set Private Members
            Me._id = _id
            Me._model = _model
            Me._floorType = _floorType
            Me._engineType = _engineType
            Me._lengthInMetres = _lengthInMetres
            Me._widthInMetres = _widthInMetres
            Me._heightInMetres = _heightInMetres
            Me._registeredPassengers = _registeredPassengers
            Me._isDoubleDecker = _doubleDecker

        End Sub

        'Construction Validators Helpers                                     
        Private Function ModelOK(ByVal model As String) As Boolean

            model = model.ToLower

            If model Is Nothing OrElse model.Trim.Length = 0 Then Return False

            Return True

        End Function

        Private Function FloorTypeOK(ByVal floorType As String) As Boolean

            floorType = floorType.ToLower

            If floorType Is Nothing OrElse floorType.Trim.Length = 0 Then Return False

            If floorType <> "raised floor" AndAlso floorType <> "low floor" AndAlso floorType <> "semi low floor" Then Return False

            Return True

        End Function

        Private Function EngineOK(ByVal engine As String) As Boolean

            engine = engine.ToLower

            If engine Is Nothing OrElse engine.Trim.Length = 0 Then Return False

            If engine <> "diesel" AndAlso engine <> "gas" AndAlso engine <> "hybrid" Then Return False

            Return True

        End Function

        Private Function DimensionOK(ByVal dimension As Double) As Boolean

            Return dimension > 0.5

        End Function

        Private Function PassengersOK(ByVal registeredPassengers As Integer) As Boolean

            Return registeredPassengers > 1


        End Function

        'To String function
        Public Overrides Function ToString() As String

            Return String.Format("{0},{1},{2},{3},{4},{5},{6},{7}", _model, _floorType, _engineType, _lengthInMetres, _widthInMetres, _heightInMetres, _registeredPassengers, _isDoubleDecker)

        End Function

    End Class


End Namespace

File: VectoCore/VectoCore/Models/BusAuxiliaries/DownstreamModules/Impl/HVAC/Bus.cs
using System;

namespace TUGraz.VectoCore.Models.BusAuxiliaries.DownstreamModules.Impl.HVAC
{
	public class Bus : IBus
	{

		// Private Property Backing
		private int _id;
		private string _model;
		private string _floorType;
		private string _engineType;
		private double _lengthInMetres;
		private double _widthInMetres;
		private double _heightInMetres;
		private int _registeredPassengers;
		private bool _isDoubleDecker;

		public int Id
		{
			get {
				return _id;
			}
		}

		public string Model
		{
			get {
				return _model;
			}
			set {
				if (!ModelOK(value))
					throw new ArgumentException("Model argument is invalid");
				_model = value;
			}
		}

		public string FloorType
		{
			get {
				return _floorType;
			}
			set {
				if (!FloorTypeOK(value))
					throw new ArgumentException("Model argument is invalid");
				_floorType = value;
			}
		}

		public string EngineType
		{
			get {
				return _engineType;
			}
			set {
				if (!EngineOK(value))
					throw new ArgumentException("EngineType argument is invalid");
				_engineType = value;
			}
		}

		public double LengthInMetres
		{
			get {
				return _lengthInMetres;
			}
			set {
				if (!DimensionOK(value))
					throw new ArgumentException("Invalid Length");
				_lengthInMetres = value;
			}
		}

		public double WidthInMetres
		{
			get {
				return _widthInMetres;
			}
			set {
				if (!DimensionOK(value))
					throw new ArgumentException("Invalid Width");
				_widthInMetres = value;
			}
		}

		public double HeightInMetres
		{
			get {
				return _heightInMetres;
			}
			set {
				if (!DimensionOK(value))
					throw new ArgumentException("Invalid Height");
				_heightInMetres = value;
			}
		}

		public int RegisteredPassengers
		{
			get {
				return _registeredPassengers;
			}
			set {
				if (!PassengersOK(value))
					throw new ArgumentException("Invalid Number Of Passengers");
				_registeredPassengers = value;
			}
		}

		public bool IsDoubleDecker
		{
			get {
				return _isDoubleDecker;
			}
			set {
				_isDoubleDecker = value;
			}
		}

		// Constructors
		public Bus(int _id, string _model, string _floorType, string _engineType, double _lengthInMetres, double _widthInMetres, double _heightInMetres, int _registeredPassengers, bool _doubleDecker)
		{


			// Validity checks.
			if (!ModelOK(_model))
				throw new ArgumentException("Model argument is invalid");
			if (!FloorTypeOK(_floorType))
				throw new ArgumentException("Model argument is invalid");
			if (!EngineOK(_engineType))
				throw new ArgumentException("EngineType argument is invalid");
			if (!DimensionOK(_lengthInMetres))
				throw new ArgumentException("Invalid Length");
			if (!DimensionOK(_widthInMetres))
				throw new ArgumentException("Invalid Width");
			if (!DimensionOK(_heightInMetres))
				throw new ArgumentException("Invalid Height");
			if (!PassengersOK(_registeredPassengers))
				throw new ArgumentException("Invalid Number Of Passengers");


			// Set Private Members
			this._id = _id;
			this._model = _model;
			this._floorType = _floorType;
			this._engineType = _engineType;
			this._lengthInMetres = _lengthInMetres;
			this._widthInMetres = _widthInMetres;
			this._heightInMetres = _heightInMetres;
			this._registeredPassengers = _registeredPassengers;
			this._isDoubleDecker = _doubleDecker;
		}

		// Construction Validators Helpers                                     
		private bool ModelOK(string model)
		{
			model = model.ToLower();

			if (model == null || model.Trim().Length == 0)
				return false;

			return true;
		}

		private bool FloorTypeOK(string floorType)
		{
			floorType = floorType.ToLower();

			if (floorType == null || floorType.Trim().Length == 0)
				return false;

			if (floorType != "raised floor" && floorType != "low floor" && floorType != "semi low floor")
				return false;

			return true;
		}

		private bool EngineOK(string engine)
		{
			engine = engine.ToLower();

			if (engine == null || engine.Trim().Length == 0)
				return false;

			if (engine != "diesel" && engine != "gas" && engine != "hybrid")
				return false;

			return true;
		}

		private bool DimensionOK(double dimension)
		{
			return dimension > 0.5;
		}

		private bool PassengersOK(int registeredPassengers)
		{
			return registeredPassengers > 1;
		}

		// To String function
		public override string ToString()
		{
			return string.Format("{0},{1},{2},{3},{4},{5},{6},{7}", _model, _floorType, _engineType, _lengthInMetres, _widthInMetres, _heightInMetres, _registeredPassengers, _isDoubleDecker);
		}
	}
}
